File: kai_universe_ui/src/shared/domain/model/ports/ModelRepository.ts
import type { Model } from '../entities/Model';
import type { ModelFormat } from '../value-objects/ModelFormat';
import type { ModelId } from '../value-objects/ModelId';

export interface ModelSearchQuery {
  readonly text?: string;
  readonly format?: ModelFormat;
  readonly compatibleOnly?: boolean;
  readonly tags?: readonly string[];
}

// Read-side port. Mutations on the catalog are out of scope for the mock.
export interface ModelRepository {
  list(): Promise<readonly Model[]>;
  search(query: ModelSearchQuery): Promise<readonly Model[]>;
  findById(id: ModelId): Promise<Model | null>;
  staffPicks(): Promise<readonly Model[]>;
}

File: kai_universe_backend/app/modules/discover/domain/repositories/model_repository.py
from abc import ABC, abstractmethod
from app.modules.discover.domain.entities.model import Model


class ModelRepository(ABC):

    @abstractmethod
    def list(self) -> list[Model]:
        """
        Return all models.
        """
        raise NotImplementedError

    @abstractmethod
    def find_by_id(self, model_id: str) -> Model | None:
        """
        Return a model by id.
        """
        raise NotImplementedError

    @abstractmethod
    def search(self, text: str | None = None, format: str | None = None, tags: list[str] | None = None,) -> list[Model]:
        """
        Search models.
        """
        raise NotImplementedError

    @abstractmethod
    def staff_picks(self) -> list[Model]:
        """
        Return staff-picked models.
        """
        raise NotImplementedError